gnosis/eth/oracles/cowswap.py

```diff
@@ -0,0 +1,74 @@
+import logging
+from typing import Optional
+
+from gnosis.protocol import GnosisProtocolAPI, OrderKind
+
+from .. import EthereumClient, EthereumNetworkNotSupported
+from .exceptions import CannotGetPriceFromOracle
+from .oracles import PriceOracle
+from .utils import get_decimals
+
+logger = logging.getLogger(__name__)
+
+
+class CowswapOracle(PriceOracle):
+    """
+    CowSwap Oracle implementation
+
+    https://docs.cow.fi/off-chain-services/api
+    """
+
+    def __init__(
+        self,
+        ethereum_client: EthereumClient,
+    ):
+        """
+        :param ethereum_client:
+        """
+        self.ethereum_client = ethereum_client
+        self.w3 = ethereum_client.w3
+        self.api = GnosisProtocolAPI(ethereum_client.get_network())
+
+    @classmethod
+    def is_available(
+        cls,
+        ethereum_client: EthereumClient,
+    ) -> bool:
+        """
+        :param ethereum_client:
+        :return: `True` if CowSwap is available for the EthereumClient provided, `False` otherwise
+        """
+        try:
+            GnosisProtocolAPI(ethereum_client.get_network())
+            return True
+        except EthereumNetworkNotSupported:
+            return False
+
+    def get_price(
+        self, token_address_1: str, token_address_2: Optional[str] = None
+    ) -> float:
+        token_address_2 = token_address_2 or self.api.weth_address
+        if token_address_1 == token_address_2:
+            return 1.0
+
+        token_1_decimals = get_decimals(token_address_1, self.ethereum_client)
+        try:
+            result = self.api.get_estimated_amount(
+                token_address_1, token_address_2, OrderKind.SELL, 10**token_1_decimals
+            )
+            if "amount" in result:
+                # Decimals needs to be adjusted
+                token_2_decimals = get_decimals(token_address_2, self.ethereum_client)
+                return float(result["amount"]) / 10**token_2_decimals
```

Comment:
Should we use `fractions` here since we are dealing with financial decimal values?

Reply:
That's a good idea, but I wouldn't do it as it would break oracle interface for all the implemented oracles. At some point we can refactor all the oracles to return `Fraction`